### DISCUSSION TITLE
External ADC + external DAC on esp32

### DISCUSSION THREAD

**Discussion by gurinkirill** [pschatzmann/arduino-audio-tools#1525] [Q&A] [answered]
I have next scheme: _mic -> DAC ([PCM5102A](https://aliexpress.ru/item/32830811600.html?sku_id=12000031502128383&spm=.search_results.1.7d4a67fe30ROpw)) -> esp32 -> ADC ([PCM1808](https://aliexpress.ru/item/1005006406148529.html?sku_id=12000037052878393&spm=a2g2x.productlist.search_results.11.188218bdbLMdt5)) -> speaker_. I configure these like:
```
#include "Arduino.h"
#include "AudioTools.h"

I2SStream in;
I2SStream out;
StreamCopy copier(out, in);

void setup() {

    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Info);

    // ADC
    auto configAdc = in.defaultConfig(RX_MODE);
    configAdc.port_no = 0;
    configAdc.pin_bck = 33;
    configAdc.pin_ws = 32;
    configAdc.pin_data = 25;
    configAdc.i2s_format = I2S_STD_FORMAT;
    configAdc.bits_per_sample = 32;
    configAdc.channels = 2;
    configAdc.sample_rate = 44100;
    configAdc.is_master = true; // optional because default setting
    configAdc.use_apll = false;
    configAdc.pin_mck = 3;
    in.begin(configAdc);

    // DAC
    auto configDac = out.defaultConfig(TX_MODE);
    configDac.port_no = 1;
    configDac.pin_bck = 14;
    configDac.pin_ws = 26;
    configDac.pin_data = 27;
    configDac.i2s_format = I2S_STD_FORMAT;
    configDac.bits_per_sample = 32;
    configDac.channels = 2;
    configDac.sample_rate = 44100;
    configDac.is_master = true;
    configDac.use_apll = false;
    configDac.pin_mck = 3;
    out.begin(configDac);
}

void loop() {
    copier.copy(); 
}
```

After uploading there is only noise in speakers. What can be the problem?

The DAC works well with internal esp32's ADC, but after connecting external one I'm facing that problem. So, I think that the problem may be related to incorrect ADC setting. I tried to change ADC's .is_master to false, nothing happens.

Connection scheme: 
![Снимок экрана 2024-04-23 111229](https://github.com/pschatzmann/arduino-audio-tools/assets/91548763/aa30b646-e2db-4d49-a218-dddfd138867d)

**Comment by pschatzmann** [comment, marked as answer]
Just look at the provided examples: There is a one port soluton where you can share the pins and a two port solution where you would need to specify the port:

- [streams-i2s-i2s-2](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-stream/streams-i2s-i2s-2)
- [streams-i2s-i2s](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-stream/streams-i2s-i2s)

**Comment by AlexSWapk** [comment]
Hello @gurinkirill, 
Did you manage to get your project working? Because I have a similar project and I don't get any sound coming out of the PCM5102 (which works fine on its own) - I think it's the PCM1808 that's causing the problem...
Many thanks in advance !

**Reply by gurinkirill** [reply]
Hi! Yes, I did. [Here is](https://github.com/gurinkirill/DSAP) the project repo. And [this is the configuration](https://github.com/gurinkirill/DSAP/blob/main/src%2Fmain.cpp#L156) of DAC and ADC, I used.

**Reply by AlexSWapk** [reply]
@gurinkirill 
Thank you for your feedback and your project! I feel like I'm eliminating problems because I have my I2S signals but I don't have any signal coming out of DATA from the PCM1808. Could you please tell me which jumper you soldered (OP1 / OP2 / OP3) corresponding to the MS OPTION?

**Reply by gurinkirill** [reply]
Unfortunately, now I don't remember all the details, but I used this PCM1808 [from ali](https://sl.aliexpress.ru/p?key=uxBE3xN). As i remember I tried to change jumper, but in the end i just set it to default state - _OP1, OP2, OP3 open_.
![изображение](https://github.com/user-attachments/assets/d36d86ad-af35-43d1-83e8-b5820af91930)

Also i remember that when i tried to use ADC's JACK with microphone there was no sound from ADC and on plotter in Arduino IDE i saw smth like noizes. After that I decided to use MAX9814 micro and this made everything work.

**Reply by gurinkirill** [reply]
Maybe problem is in connection? Some esp's pins are inappropriate for i2s communication

**Reply by AlexSWapk** [reply]
Thank you, I will try this connection of the jumpers !
 But did you still keep the pcm1808 with the microphone ? Or do you have your microphone directly connected to the ESP's ADC ?

**Reply by gurinkirill** [reply]
Mic out is connected to ADC input, mic voltage and ground to esp. MAX9814 (mic) _out_ pin → PCM1808 ADC _R_IN_

**Reply by AlexSWapk** [reply]
Thank you so much for taking the time to help me! I'll try this when I have time.

**Reply by gurinkirill** [reply]
If you have any more questions, you are welcome)
